src/irr/asset/CGLSLBRDFBuiltinIncludeLoader.h

```diff
@@ -426,6 +465,43 @@ float irr_glsl_ggx_burley_aniso(float anisotropy, float a2, float TdotH, float B
 	return w2*w2*atab * irr_glsl_RECIPROCAL_PI;
 }
 
+//accounting for visibility of microfacets so basically this one should always be used because converges faster and doesnt "lose samples"
+//https://schuttejoe.github.io/post/ggximportancesamplingpart2/
+//again: everything happens in reflection space, so _V is supposed to be view vector in reflection space (geom normal is positive Z axis)
+//https://hal.archives-ouvertes.fr/hal-01509746/document
+//Also: problem is our anisotropic ggx ndf (above) has extremely weird API (anisotropy and a2 instead of ax and ay) and so it's incosistent with sampling function
+//Note: this is kinda blurry and i dont really understand whats going on here
+BSDFSample irr_glsl_ggx_cos_gen_sample(in uvec2 _sample, in vec3 _V, in float _ax, in float _ay)
+{
+    vec2 u = vec2(_sample)/float(4294967295);
+
+    // stretch view vector so that we're sampling as if roughness=1.0
+    vec3 V = normalize(vec3(_ax*_V.x, _ay*_V.y, _V.z));
```

Comment:
you need to use `vec3(TdotV,BdotV,NdotV)` instead of `V` this whole thing only works with a tangent space where N=(0,0,1)

Reply:
also need to handle a special case when length of stretched vector is close to zero (smooth surface and grazing angle), or this whole thing will be full of `NaN` (when `NdotV` is close to zero, we have a tendency of `a--->1.0`, etc.)

Meaning you can just skip the whole body and set `TdotH=0.0;BdotH=0.0;NdotH=1.0;`

Reply:
actually you should keep track of the sign of `NdotV` because you should use a temporary var equal to `abs(NdotV)`, then do the whole sampling, then at the end make the sign of `NdotH` match `NdotV`